File: exercise/LeetCode/3. Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.cpp

#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) return 0;

        std::vector<int> pos(256, -1);
        int maxlen = 1;
        int ss = 0; // 子串的起始下标
        for (int k = 0; k < s.size(); ++k)
        {
            // 如果pos中存在s[k]，则ss指向已存在的s[k]的下一个字符
            if (pos[s[k]] >= 0)
                ss = std::max(ss, pos[s[k]]+1);
            // 记录字符s[k]
            pos[s[k]] = k;
            maxlen = std::max(maxlen, k + 1 - ss);
        }
        return maxlen;
    }
};


int main()
{
    Solution su;
    cout << su.lengthOfLongestSubstring("dvdfad");
    return 0;
}

File: exercise/LeetCode/3. Longest Substring Without Repeating Characters/Longest Substring Without Repeating Characters.py

class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == "":
            return 0
        mlen = 1
        ss = 0
        pos = {}
        for idx,ch in enumerate(s):
            if ch in pos:
                ss = max(ss, pos[ch] + 1)
            pos[ch] = idx
            mlen = max(mlen, idx + 1 - ss)
        return mlen

if __name__ == "__main__":
    su = Solution()
    print(su.lengthOfLongestSubstring("abba"))
